auxx/src/Lang/Interpreter.hs

```diff
@@ -0,0 +1,75 @@
+{-# LANGUAGE RankNTypes #-}
+
+module Lang.Interpreter
+       ( Value(..)
+       , evaluate
+       , EvalError(..)
+       ) where
+
+import           Universum
+
+import           Control.Monad.Except (throwError)
+
+import           Lang.Argument        (ProcError, consumeArguments)
+import           Lang.Command         (CommandProc (..))
+import           Lang.Name            (Name)
+import           Lang.Syntax          (Expr (..), Lit (..), ProcCall (..))
+import           Lang.Value           (Value (..))
+
+data EvalCtx m = EvalCtx
+    { ecCommandProcs :: [CommandProc m]
+    } deriving ()
+
+data EvalError
+    = CommandNotSupported Name
+    | InvalidArguments Name ProcError
+    deriving (Eq, Ord, Show)
+
+instance Exception EvalError
+
+type T m a = Monad m =>
+    ReaderT (EvalCtx m) (ExceptT EvalError m) a
+
+evaluate :: Monad m => [CommandProc m] -> Expr -> m (Either EvalError Value)
+evaluate commandProcs expr = runExceptT $ runReaderT (eval expr) ctx
+  where
+    ctx = EvalCtx { ecCommandProcs = commandProcs }
+
+eval :: Expr -> T m Value
+eval = \case
+    ExprUnit -> return ValueUnit
+    ExprLit l -> return (literalToValue l)
+    ExprGroup exprs -> evalExprGroup exprs
+    ExprProcCall procCall ->
+        evalProcCall =<< traverse eval procCall
+
+evalExprGroup :: NonEmpty Expr -> T m Value
+evalExprGroup (x :| xs) = case nonEmpty xs of
```

Comment:
Is it equivalent to `NE.last . traverse eval`?

Reply:
yes but it's single-pass. I'll add a comment